### ISSUE TITLE
Bound typevar being expanded to union of constraints

### ISSUE THREAD

**Issue by d-tw** [microsoft/pyright#1911]
**Describe the bug**
In the code snippet below, the instance variable `value` is of type `Optional[T]`, where `T` is defined as `TypeVar('T', str, int, float, bool)`, yet after comparing `value` to `None`, the type of `value` becomes `str | int | float | bool`, instead of `T`.

**To Reproduce**
See screenshot/code below.

**Expected behavior**
After the `f self.value is None:` condition, the type of `self.value` should be narrowed from `Optional[T]` to `T`.

**Screenshots or Code**
```python
from typing import Generic, Optional, TypeVar

T = TypeVar('T', str, int, float, bool)

class Item(Generic[T]):
    value: Optional[T]

    def __init__(self, source: Optional[T]) -> None:
        self.value = source

    def read(self) -> Optional[T]:        
        reveal_type(self.value)

        if self.value is None:
            raise RuntimeError(f"Item is required!")

        reveal_type(self.value)

        return self.value
```

<img width="993" alt="image" src="https://user-images.githubusercontent.com/3317531/119708499-83a79a00-be5c-11eb-8db3-8b2b1cd26ae1.png">

```sh
Found 1 source file
../sample.py
  ../sample.py:18:21 - info: Type of "self.value" is "Source[T@Item]"
  ../sample.py:23:21 - info: Type of "self.value" is "T@Item | str | int | float | bool | None"
  ../sample.py:25:16 - error: Expression of type "T@Item | str | int | float | bool | None" cannot be assigned to return type "T@Item | None"
    Type "T@Item | str | int | float | bool | None" cannot be assigned to type "T@Item | None"
      Type "str" cannot be assigned to type "T@Item | None"
        Type "str" cannot be assigned to type "T@Item"
        Cannot assign to "None"
      Type "int" cannot be assigned to type "T@Item | None"
        Type "int" cannot be assigned to type "T@Item"
        Cannot assign to "None"
      Type "float" cannot be assigned to type "T@Item | None"
    ... (reportGeneralTypeIssues)
1 error, 0 warnings, 2 infos 
Completed in 0.577sec
```

**VS Code extension or command-line**
pyright 1.1.143

**Comment by erictraut** [comment]
Thanks for the bug report. This prompted me to do a long-overdue overhaul of the way the type evaluator tracks [conditional types](https://github.com/microsoft/pyright/blob/main/docs/type-concepts.md#constrained-type-variables-and-conditional-types).

This will be fixed in the next release.

**Comment by erictraut** [comment]
This is included in pyright 1.1.145, which I just published. It will also be included in the next release of pylance.
